File: 637-average-of-levels-in-binary-tree/637-average-of-levels-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        self.res = []
        self.levelNum = []
        self.reachedLevel = -1
        self.traverse(root, 0)
        self.calculateAve()
        return self.res
        
    def traverse(self, root, level):
        if self.reachedLevel < level:
            self.reachedLevel += 1
            self.res.append(0)
            self.levelNum.append(0)
        self.res[level] += root.val
        self.levelNum[level] += 1
        if root.left is not None:
            self.traverse(root.left, level + 1)
        if root.right is not None:
            self.traverse(root.right, level + 1)
            
    def calculateAve(self):
        for i in range(len(self.res)):
            self.res[i] = round(self.res[i] / self.levelNum[i], 5)

File: 637-average-of-levels-in-binary-tree/637-average-of-levels-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public HashMap<Integer, Double> sum = new HashMap<>();
    public HashMap<Integer, Integer> count = new HashMap<>();
    public void traverse(TreeNode root, int level){
        if(root == null) return;
        sum.put(level, sum.getOrDefault(level, 0.0) + root.val);
        count.put(level, count.getOrDefault(level, 0) + 1);
        traverse(root.left, level + 1);
        traverse(root.right, level + 1);

    }
    public List<Double> averageOfLevels(TreeNode root) {
        List<Double> res = new LinkedList<>();
        traverse(root, 0);
        int i = 0;
        while(sum.containsKey(i)){
            res.add((double)sum.get(i) / (double)count.get(i));
            i++;
        }
        return res;
    }
}